### ISSUE TITLE
cp_height protocol support

### ISSUE THREAD

**Issue by JeremyRand** [spesmilo/electrum#4900]
Version 1.4 of the ElectrumX protocol added a `cp_height` argument to the `blockchain.block.header` and `blockchain.block.headers` methods, which allows (AFAICT) more compact proofs that a header is committed to by a checkpoint.  See https://github.com/kyuupichan/electrumx/blob/master/docs/protocol-methods.rst#blockchainblockheader for an example.

AFAICT Electrum doesn't utilize this feature at all right now.  Would Electrum be interested in adding support for this?

(The amount of data currently transferred when downloading headers in Namecoin's Electrum port is prohibitively large for some applications, and I suspect that any Namecoin-specific fixes that I might implement would trigger merge conflicts if Electrum added `cp_height` support, so I figure I should reach out to you guys rather than just doing my own thing.)

**Comment by JeremyRand** [comment]
Right, so the current checkpointing mechanism has 2 issues for Namecoin:

1. Namecoin's fix for the timewarp attack (which Bitcoin didn't adopt because it would be a hardfork) changes which block heights are used for retargeting, so the target data in Electrum's checkpoints is for the wrong block heights when applied to Namecoin.
2. Electrum's mechanism needs to download all of the headers that connect a desired block height to the height of a checkpoint.  In Namecoin's case, this often is several megabytes (due to the larger size of Namecoin's headers because of the AuxPoW data).  Namecoin's name lookup functionality needs relatively low latency, and needing to download several megabytes when doing a DNS lookup is prohibitive.

I was able to work around issue 1 with a stupid 1-liner hack that's probably a bad idea but seems to produce the correct results.  Issue 2 is the one that `cp_height` would be particularly helpful for solving.  I can probably do a substantial optimization by modifying ElectrumX and Electrum-NMC to not transfer AuxPoW data when the desired block heights are behind a checkpoint (since AuxPoW data isn't needed in such cases), but it would definitely still be nice not to need to download all the headers when all we need is one header plus a Merkle proof.

I might be interested in implementing `cp_height` in Electrum (or I might inquire whether the Electron-Cash guys are interested in submitting a PR to you), but it's not the highest priority thing for me at the moment, so not sure when I'd get to it.

**Comment by JeremyRand** [comment]
It looks like https://github.com/Electron-Cash/Electron-Cash/pull/809 is the PR that added this feature to Electron-Cash.

**Comment by A60AB5450353F40E** [comment]
Did y'all know that Bitcoin's Merkle tree **contains** a "Merkle Mountain Range" ([Todd, 2012](https://github.com/opentimestamps/opentimestamps-server/blob/master/doc/merkle-mountain-range.md)) accumulator? This same MMR structure is called a "Balanced Merkle forest" in [Utreexo paper](https://eprint.iacr.org/2019/611).

This means that a client that ships with a trusted checkpoint can independently update its root to latest checkpoint and then prune even the recent headers!
Accumulator storage is O(log(n)) and appending a leaf is O(1)!

1. Starting from some old root
2. Query server for proof for the last committed leaf (`cp_height`=`height`)
3. Extract MMR "peaks" from the proof, verify against trusted root, and construct the accumulator state
4. Extend the MMR accumulator with new headers
5. Compute new root from new MMR peaks
6. Prune the headers, because they're now covered by the updated root

Which proof elements are the "peaks"?

Observe the structure:

                                 **root**
                         __________/  \________
                        /                      \
                      (H)                       c
                   _ /   \ _                   / \
                 /           \                /   \
                F             G              b     b'
              /    \        /    \        /     \
             A      B      C      D     (E)      a
            / \    / \    / \    / \    / \     /  \
           0   1  2   3  4   5  6   7  8   9  (10)  10'

    Note: nodes marked with ' are duplicates; lowercase letters mark the auxiliary subtree; () mark the MMR peaks.

The proof will be: 10', (E), b', (H)

We need: (10), (E), (H) but (10)=10' so we get it from the server.

Proof of concept implementation and tests: https://github.com/cculianu/Fulcrum/issues/187#issuecomment-3975472620

---

There's a more efficient Merkle tree construction than Bitcoin's: instead of duplicating the nodes we could iteratively construct an MMR on top of previous MMR's peaks, until the last MMR only has 1 peak:

    Iterative MMR Construction

    Second bag MMR:                   **root**
                                    ____/  \____
                                   /            \
    First bag MMR:             ((I))            |
                         _______/ \______       |
                        /                \      |
    Base MMR:         (H)                |      |
                   _ /   \ _             |      |
                 /           \           |      |
                F             G          |      |
              /    \        /    \       |      |
             A      B      C      D     (E)     |
            / \    / \    / \    / \    / \     |
           0   1  2   3  4   5  6   7  8   9  ((10))

     ( )  = peaks of the base MMR (leaves of the first bag MMR)
    (( )) = peaks of the first bag MMR
    root  = only peak of the second bag MMR

Another way to view this construction is as rightward promotion: to pair the highest peak, we promote the next peak to its level; to pair that result, we promote the next peak, and so on.

This results in smaller proofs and lower number of merges when appending. I wonder if it would be worth the effort to implement this structure in the future.

Main benefit: smallest proofs for more recent headers.
